cachi2/core/models/sbom.py

```diff
@@ -93,3 +97,473 @@ class Sbom(pydantic.BaseModel):
+                        )
+                    ],
+                    annotations=annotations,
+                )
+            )
+        for package in packages:
+            relationships.append(
+                SPDXRelation(
+                    spdxElementId="SPDXRef-DOCUMENT",
+                    comment="",
+                    relatedSpdxElement=package.SPDXID,
+                    relationshipType="CONTAINS",
+                )
+            )
+        return SPDXSbom(
+            packages=packages,
+            relationships=relationships,
+            creationInfo=SPDXCreationInfo(
+                creators=[f"{tool.vendor} {tool.name}" for tool in self.metadata.tools]
+            ),
+        )
+
+
+class SPDXPackageExternalRefReferenceLocatorURI(pydantic.BaseModel):
+    """SPDX Package External Reference with URI reference locator."""
+
+    referenceLocator: str
+
+    @pydantic.validator("referenceLocator")
+    @classmethod
+    def _validate_uri_reference_locator(cls, referenceLocator: str) -> str:
+        parsed = urlparse(referenceLocator)
+        if not (parsed.scheme and (parsed.path or parsed.netloc)):
+            raise ValueError("Invalid URI reference locator")
+        return referenceLocator
+
+
+class SPDXPackageExternalRef(pydantic.BaseModel):
+    """SPDX Package External Reference.
+
+    Compliant to the SPDX specification:
+    https://spdx.github.io/spdx-spec/v2.3/package-information/#721-external-reference-field
+    """
+
+    referenceLocator: str
+    referenceType: str
+    referenceCategory: str
+
+    def __hash__(self) -> int:
+        return hash((self.referenceLocator, self.referenceType, self.referenceCategory))
+
+
+class SPDXPackageExternalRefSecurity(SPDXPackageExternalRef):
+    """SPDX Package External Reference for security category.
+
+    Compliant to the SPDX specification:
+    https://spdx.github.io/spdx-spec/v2.3/package-information/#721-external-reference-field
+    """
+
+    referenceCategory: Literal["SECURITY"]
+
+
+class SPDXPackageExternalRefSecurityCPE22(SPDXPackageExternalRefSecurity):
+    """SPDX Package External Reference for security category and type cpe22.
+
+    Compliant to the SPDX specification:
+    https://spdx.github.io/spdx-spec/v2.3/package-information/#721-external-reference-field
+    """
+
+    referenceCategory: Literal["SECURITY"]
+    referenceLocator: str = pydantic.Field(
+        pattern=r"[c][pP][eE]:/[AHOaho]?(:[A-Za-z0-9\._\-~%]*){0,6}"
+    )
+    referenceType: Literal["cpe22Type"]
+
+
+class SPDXPackageExternalRefSecurityCPE23(SPDXPackageExternalRefSecurity):
+    """SPDX Package External Reference for security category and type cpe23.
+
+    Compliant to the SPDX specification:
+    https://spdx.github.io/spdx-spec/v2.3/package-information/#721-external-reference-field
+    """
+
+    referenceCategory: Literal["SECURITY"]
+    referenceLocator: str = pydantic.Field(
+        pattern=r"""cpe:2\.3:(?P<part>[aho\*\-])
```

Comment:
Thank you for adding group names! This makes it much easier to follow. Please add a comment stating that this regexp essentially replicates one from MITRE and add a link to source.

Reply:
Awesome, thank you!